src/parse_analyze.c

```diff
@@ -0,0 +1,121 @@
+#include <postgres.h>
+#include <nodes/nodeFuncs.h>
+#include <parser/analyze.h>
+
+#include "cache.h"
+#include "hypertable.h"
+#include "hypertable_cache.h"
+#include "extension.h"
+#include "parse_rewrite.h"
+
+void		_parse_analyze_init(void);
+void		_parse_analyze_fini(void);
+
+static post_parse_analyze_hook_type prev_post_parse_analyze_hook = NULL;
+
+typedef struct HypertableQueryCtx
+{
+	Query	   *parse;
+	Query	   *parent;
+	CmdType		cmdtype;
+	Cache	   *hcache;
+	Hypertable *hentry;
+} HypertableQueryCtx;
+
+/*
+ * Identify queries on a hypertable by walking the query tree. If the query is
+ * indeed on a hypertable, setup the necessary state and/or make modifications
+ * to the query tree.
+ */
+static bool
+hypertable_query_walker(Node *node, void *context)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) node;
+		HypertableQueryCtx *ctx = (HypertableQueryCtx *) context;
+
+		if (rte->rtekind == RTE_RELATION)
+		{
+			Hypertable *hentry = hypertable_cache_get_entry(ctx->hcache, rte->relid);
+
+			if (hentry != NULL)
+				ctx->hentry = hentry;
+		}
+
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		bool		result;
+		HypertableQueryCtx *ctx = (HypertableQueryCtx *) context;
+		CmdType		old = ctx->cmdtype;
+		Query	   *query = (Query *) node;
+		Query	   *oldparent = ctx->parent;
+
+		/* adjust context */
+		ctx->cmdtype = query->commandType;
+		ctx->parent = query;
+
+		result = query_tree_walker(ctx->parent, hypertable_query_walker,
+								   context, QTW_EXAMINE_RTES);
+
+		/* restore context */
+		ctx->cmdtype = old;
+		ctx->parent = oldparent;
+
+		return result;
+	}
+
+	return expression_tree_walker(node, hypertable_query_walker, context);
+}
+
+static void
+timescaledb_post_parse_analyze(ParseState *pstate, Query *query)
+{
+	if (NULL != prev_post_parse_analyze_hook)
+		/* Call any earlier hooks */
+		prev_post_parse_analyze_hook(pstate, query);
+
+	if (extension_is_loaded())
+	{
+		HypertableQueryCtx context = {
+			.parse = query,
+			.parent = query,
+			.cmdtype = query->commandType,
+			.hcache = hypertable_cache_pin(),
+			.hentry = NULL,
+		};
+
+		/* The query for explains is in the utility statement */
+		if (query->commandType == CMD_UTILITY &&
+			IsA(query->utilityStmt, ExplainStmt))
+			query = (Query *) ((ExplainStmt *) query->utilityStmt)->query;
+
+		hypertable_query_walker((Node *) query, &context);
+
+		/* note assumes 1 hypertable per query */
+		if (NULL != context.hentry)
+			parse_rewrite_query(pstate, query, context.hentry);
+
+		cache_release(context.hcache);
+	}
+}
+
+void
+_parse_analyze_init(void)
+{
+	prev_post_parse_analyze_hook = post_parse_analyze_hook;
```

Comment:
Three reasons:

1. It made more sense to me that parse/rewrite-related stuff is handled at the "parsing" stage rather than "planning" stage, since none of the stuff we are doing in this hook is related to planning. Imagine, for instance, that another extension inserts itself before our planning hook, then that extension should, IMO, see our parsing-related rewrites so that it can run its stuff on the final query.
2. It is cleaner to move this code into a separate source file as planner.c was pretty massive.
3. Parse-related functions (like `ParseFuncOrColumn()`) requires access to the `ParseState` that is given in the `post_parse_analyze` hook and PG10 also requires access to `pstate->p_last_srf`: ```last_srf should be a copy of pstate->p_last_srf from just before we started transforming fargs.  If the caller knows that fargs couldn't contain any SRF calls, last_srf can just be pstate->p_last_srf.``` Weirdly, though, PG96 also requires passing the `ParseState`, but this was always set to `NULL` in the code for reasons I don't know. When looking at the code it certainly seemed like this could cause a NULL-pointer error in some cases, although, incidently, it obviously didn't happen for most of our use cases. Maybe you can better answer why it is valid to pass a `NULL` ParseState to these functions?

One thing I could do is to move this hook change into a separate commit (or even separate PR) if it makes sense to you.

Reply:
Let's keep in this PR. we just have to be mindful not to merge multi-version stuff before fixing this.